File: benchmarks/matty_suite/bench_fft_1m.m
% bench_fft_1m.m -- MATLAB port of bench_fft_1m.qu.
%
% Same exact-bin-sinusoid ground truth and phase convention, same
% methodology. fft/ifft are base MATLAB (confirmed already working in
% bench_fair.m earlier this session) -- unlike bench_filter_accuracy.m,
% this needs no toolbox.

run_case(1048576, 'power-of-two  ');
run_case(1000000, 'non-power-of-2');

function run_case(N, label)
    sig = make_signal(N);

    X = fft(sig.x); %#ok<NASGU> warm-up, discarded
    best_t = inf;
    xr = [];
    for r = 1:3
        tic;
        X = fft(sig.x);
        xr = ifft(X);
        t = toc;
        best_t = min(best_t, t);
    end

    [max_signal_err, max_noise] = check_accuracy(X, sig, N);
    roundtrip_err = max(abs(xr - sig.x));

    fprintf('%s (N=%d): time=%.6f s   max signal-bin error=%.2e   noise floor=%.2e   roundtrip |ifft(fft(x))-x|=%.2e\n', ...
        label, N, best_t, max_signal_err, max_noise, roundtrip_err);
end

function sig = make_signal(N)
    k1 = 137; A1 = 1.0; phi1 = 0.0;
    k2 = 9973; A2 = 2.5; phi2 = pi / 3.0;
    k3 = round(N / 5.0) + 11; A3 = 0.7; phi3 = -pi / 4.0;

    n = 0:(N-1);
    x = A1 * cos(2.0 * pi * k1 * n / N) + A2 * cos(2.0 * pi * k2 * n / N + phi2) + A3 * cos(2.0 * pi * k3 * n / N + phi3);

    sig.x = x; sig.k1 = k1; sig.A1 = A1; sig.phi1 = phi1;
    sig.k2 = k2; sig.A2 = A2; sig.phi2 = phi2;
    sig.k3 = k3; sig.A3 = A3; sig.phi3 = phi3;
end

function [max_signal_err, max_noise] = check_accuracy(X, sig, N)
    % MATLAB fft() is 1-indexed: bin k (0-based) is X(k+1).
    bins0 = [sig.k1, N - sig.k1, sig.k2, N - sig.k2, sig.k3, N - sig.k3];
    amps = [sig.A1, sig.A1, sig.A2, sig.A2, sig.A3, sig.A3];
    phis = [sig.phi1, -sig.phi1, sig.phi2, -sig.phi2, sig.phi3, -sig.phi3];

    max_signal_err = 0.0;
    for i = 1:6
        k0 = bins0(i);
        expected = (N * amps(i) / 2.0) * (cos(phis(i)) + 1i * sin(phis(i)));
        err = abs(X(k0 + 1) - expected);
        max_signal_err = max(max_signal_err, err);
    end

    max_noise = 0.0;
    for k0 = 0:997:(N-1)
        if ~any(bins0 == k0)
            m = abs(X(k0 + 1));
            max_noise = max(max_noise, m);
        end
    end
end

File: benchmarks/matty_suite/bench_fft_1m.py
"""bench_fft_1m.py -- numpy port of bench_fft_1m.qu.

Same exact-bin-sinusoid ground truth, same X[k] = (N*A/2)*exp(+i*phi)
formula (phase sign verified against a real FFT at N=1024 before this
file was written -- the naive guess had it backwards), same data-
outside-the-clock / one-warm-up-discarded / best-of-3 methodology.
"""

import time

import numpy as np


def make_signal(N):
    k1, A1, phi1 = 137, 1.0, 0.0
    k2, A2, phi2 = 9973, 2.5, np.pi / 3.0
    k3, A3, phi3 = round(N / 5.0) + 11, 0.7, -np.pi / 4.0

    n = np.arange(N, dtype=float)
    x = (
        A1 * np.cos(2.0 * np.pi * k1 * n / N)
        + A2 * np.cos(2.0 * np.pi * k2 * n / N + phi2)
        + A3 * np.cos(2.0 * np.pi * k3 * n / N + phi3)
    )
    return dict(x=x, k1=k1, A1=A1, phi1=phi1, k2=k2, A2=A2, phi2=phi2, k3=k3, A3=A3, phi3=phi3)


def check_accuracy(X, sig, N):
    bins = [sig["k1"], N - sig["k1"], sig["k2"], N - sig["k2"], sig["k3"], N - sig["k3"]]
    amps = [sig["A1"], sig["A1"], sig["A2"], sig["A2"], sig["A3"], sig["A3"]]
    phis = [sig["phi1"], -sig["phi1"], sig["phi2"], -sig["phi2"], sig["phi3"], -sig["phi3"]]

    max_signal_err = 0.0
    for k, a, phi in zip(bins, amps, phis):
        expected = (N * a / 2.0) * (np.cos(phi) + 1j * np.sin(phi))
        err = abs(X[k] - expected)
        max_signal_err = max(max_signal_err, err)

    max_noise = 0.0
    bin_set = set(bins)
    for k in range(0, N, 997):
        if k not in bin_set:
            max_noise = max(max_noise, abs(X[k]))

    return max_signal_err, max_noise


def run_case(N, label):
    sig = make_signal(N)

    X = np.fft.fft(sig["x"])  # warm-up, discarded
    best_t = float("inf")
    xr = None
    for _ in range(3):
        t0 = time.perf_counter()
        X = np.fft.fft(sig["x"])
        xr = np.fft.ifft(X)
        t = time.perf_counter() - t0
        best_t = min(best_t, t)

    max_signal_err, max_noise = check_accuracy(X, sig, N)
    roundtrip_err = np.max(np.abs(xr - sig["x"]))

    print(
        f"{label} (N={N}): time={best_t:.6f} s   "
        f"max signal-bin error={max_signal_err:.2e}   "
        f"noise floor={max_noise:.2e}   "
        f"roundtrip |ifft(fft(x))-x|={roundtrip_err:.2e}"
    )


run_case(1048576, "power-of-two  ")
run_case(1000000, "non-power-of-2")
